sync_diff_inspector/checkpoints/checkpoints.go

```diff
@@ -0,0 +1,263 @@
+
+import (
+	"container/heap"
+	"context"
+	"encoding/json"
+	"os"
+	"sync"
+
+	"github.com/pingcap/tidb-tools/sync_diff_inspector/config"
+	"github.com/pingcap/tidb-tools/sync_diff_inspector/report"
+
+	"github.com/pingcap/tidb-tools/sync_diff_inspector/chunk"
+	"github.com/siddontang/go/ioutil2"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+)
+
+var (
+
+	// SuccessState
+	// for chunk: means this chunk's data is equal
+	// for table: means this all chunk in this table is equal(except ignore chunk)
+	SuccessState = "success"
+
+	// FailedState
+	// for chunk: means this chunk's data is not equal
+	// for table: means some chunks' data is not equal or some chunk check failed in this table
+	FailedState = "failed"
+
+	// for chunk: means meet error when check, don't know the chunk's data is equal or not equal
+	// for table: don't have this state
+	errorState = "error"
+
+	// for chunk: means this chunk is not in check
+	// for table: this table is checking but not finished
+	notCheckedState = "not_checked"
+
+	// for chunk: means this chunk is checking
+	// for table: don't have this state
+	checkingState = "checking"
+
+	// for chunk: this chunk is ignored. if sample is not 100% or it is Empty chunk, will ignore some chunk
+	// for table: don't have this state
+	IgnoreState = "ignore"
+)
+
+type Node struct {
+	State string `json:"state"` // indicate the state ("success" or "failed") of the chunk
+
+	ChunkRange *chunk.Range `json:"chunk-range"`
+	IndexID    int64        `json:"index-id"`
+}
+
+func (n *Node) GetID() *chunk.ChunkID { return n.ChunkRange.Index }
+
+func (n *Node) GetState() string { return n.State }
+
+func (n *Node) GetTableIndex() int { return n.ChunkRange.Index.TableIndex }
+
+func (n *Node) GetBucketIndexLeft() int { return n.ChunkRange.Index.BucketIndexLeft }
+
+func (n *Node) GetBucketIndexRight() int { return n.ChunkRange.Index.BucketIndexRight }
+
+func (n *Node) GetChunkIndex() int { return n.ChunkRange.Index.ChunkIndex }
+
+// IsAdjacent represents whether the next node is adjacent node.
+// it's the important logic for checkpoint update.
+// we need keep this node save to checkpoint in global order.
+func (n *Node) IsAdjacent(next *Node) bool {
+	if n.GetTableIndex() == next.GetTableIndex()-1 {
+		if n.ChunkRange.IsLastChunkForTable() && next.ChunkRange.IsFirstChunkForTable() {
+			return true
+		}
+	}
+	if n.GetTableIndex() == next.GetTableIndex() {
+		// same table
+		if n.GetBucketIndexRight() == next.GetBucketIndexLeft()-1 {
+			if n.ChunkRange.IsLastChunkForBucket() && next.ChunkRange.IsFirstChunkForBucket() {
+				return true
+			}
+			return false
+		}
+		if n.GetBucketIndexLeft() == next.GetBucketIndexLeft() {
+			return n.GetChunkIndex() == next.GetChunkIndex()-1
+		}
+		return false
+	}
+	return false
+}
+
+// IsLess represents whether the cur node is less than next node.
+func (n *Node) IsLess(next *Node) bool {
+	if n.GetTableIndex() <= next.GetTableIndex()-1 {
```

Comment:
seems we could use `n.GetTableIndex() < next.GetTableIndex()` like line 117

Reply:
Done.